orttraining/orttraining/core/graph/graph_parser/pattern_graph.h

```diff
@@ -0,0 +1,440 @@
+#pragma once
+
+#include "core/graph/contrib_ops/onnx_function_util.h"
+#include "orttraining/core/graph/graph_augmenter.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/model.h"
+
+#define PARSER_LOG \
+  LOGS_DEFAULT(WARNING)
+
+namespace onnxruntime {
+namespace training {
+
+// First: graph node index. Second: corresponding pattern node.
+typedef std::pair<onnxruntime::NodeIndex, const onnxruntime::Node*> PNN;
```

Comment:
>std::pair<onnxruntime::NodeIndex, const onnxruntime::Node*> PNN; [](http://example.com/codeflow?start=8&length=64)

This is the result we returned in a successfully match. Shall we return a map of "Name in the Pattern Graph" : "Matched NodeInfo"?  This will make it easier for subsquent manipulation.